### PR TITLE
[FSDP][Replicate] tests replicate gradient accumulation and 1f1b microbatching

### PR DESCRIPTION
**Pull request by anshul-si**
**Summary:** In order to ensure that replicate acts as intended (a specialized version of hsdp) we need to make sure that it can pass the same tests that fully_shard can for training. The first test verifies Replicate works with gradient accumulation properly. The second verifies that replicate works correctly with a One-Forward-One-Backward (1F1B) pipeline parallelism schedule

**Test Cases**
1. pytest test/distributed/_composable/test_replicate_training.py -k test_gradient_accumulation
2. pytest test/distributed/_composable/test_replicate_training.py -k test_1f1b_microbatching

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #163888
* #162861
* #162855
* #162853
* #162851
* __->__ #162839
* #162836
* #162830

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @ezyang @msaroufim @dcci

### CODE REVIEW

File: test/distributed/_composable/test_replicate_training.py [hunk 1]
```diff
@@ -842,5 +845,238 @@ def _test_train_shared_params(
...
+        ):
+            if reduce_scatter_only:
+                module.set_requires_all_reduce(is_last_microbatch, recurse=recurse)
+            else:
+                module.set_requires_gradient_sync(is_last_microbatch, recurse=recurse)
+
+        def set_backward_flags(_model: nn.Module, is_last_microbatch: bool):
+            if mode == "all":
+                set_grad_sync_flag(_model, is_last_microbatch)
+                if not reshard_after_backward:
```
Comment (mori360) [inline test/distributed/_composable/test_replicate_training.py:941]: does replicate have `reshard_after_backward`?

Comment (anshul-si) [reply test/distributed/_composable/test_replicate_training.py:941]: for now it does. when all functionality is complete and verified to be working fine, I will eliminate all fsdp features that are useless to replicate.

**Review by mori360 — APPROVED** [review]
